eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+    if -1 < i:
+        # linear search succeeded
+        return estimator.get_layer(index=i)
+    else:
+        raise ValueError('Could not find a suitable target layer automatically.')
+
+
+def is_suitable_activation_layer(estimator, i):
+    """
+    Check whether
+    the layer at index ``i`` matches what is required 
+    by ``estimator``.
+    
+    Matching is done by
+    * checking the rank of the layer.
+
+    Returns
+    -------
+    is suitable : boolean
+        whether the layer matches what is needed
+    """
+    # TODO: experiment with this, using many models and images, to find what works best
+    # Some ideas: 
+    # check layer type, i.e.: isinstance(l, keras.layers.Conv2D)
+    # check layer name
+    l = estimator.get_layer(index=i)
+    # a check that asks 'can we resize this activation layer over the image?'
+    rank = len(l.output_shape)
+    required_rank = len(estimator.input_shape)
+    return rank == required_rank
+
+
+def get_target_prediction(model, x, targets):
+    """
+    Get a prediction ID from ``targets``.
+
+    See documentation of ``explain_prediction_keras`` for explanation of ``targets``.
+    
+    Returns
+    -------
+    prediction id : int
+    """
+    # TODO: take in a single target as well, not just a list, 
+    # consider changing signature / types for explain_prediction generic function
+    # TODO: need to find a way to show the label for the passed prediction 
+    # as well as its probability
+
+    # TODO: maybe do the sum / loss in this function instead of grad_cam. Return a tensor.
+    # This would be consistent with what is done in https://github.com/ramprs/grad-cam/blob/master/misc/utils.lua
+    # https://github.com/ramprs/grad-cam/blob/master/classification.lua
+    # https://github.com/torch/nn/blob/master/doc/module.md
+    if isinstance(targets, list):
+        # take the first prediction from the list
+        predicted_idx = targets[0]
+        # TODO: use all predictions in the list
+        # TODO: validate list contents
+    elif targets is None:
+        predictions = model.predict(x)
+        predicted_idx = np.argmax(predictions)
+    else:
+        raise ValueError('Invalid argument "targets" (must be list or None): %s' % targets)
+    return predicted_idx
+
+
+def grad_cam(estimator, image, prediction_index, activation_layer):
+    """
+    Generate a heatmap using Gradient-weighted Class Activation Mapping (Grad-CAM).
+
+    Credits for implementation
+    * Jacob Gildenblat for "https://github.com/jacobgil/keras-grad-cam".
+    * Author of "https://github.com/PowerOfCreation/keras-grad-cam" for fixes to Jacob's implementation.
+    * Kotikalapudi, Raghavendra and contributors for "https://github.com/raghakot/keras-vis".
+    
+    Returns
+    -------
+    heatmap : object
+        A numpy.ndarray localization map.
+    """
+    # FIXME: this assumes that we are doing classification
+    # FIXME: we also explicitly assume that we are dealing with images
```

Comment:
I think both are fine for now, until we add more features.

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/f82955188f63a649ad29845763af17079237b11d